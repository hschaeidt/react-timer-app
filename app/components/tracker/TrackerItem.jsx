import React, { Component, PropTypes } from 'react';
import moment from 'moment';
import ItemControls from './item/ItemControls';

class TrackerItem extends Component {
  constructor(props) {
    super(props);

    this.state = {
      totalDuration: moment.duration(),
    };

    this.calculateTotalDuration = this.calculateTotalDuration.bind(this);
    this.handleComplete = this.handleComplete.bind(this);
    this.handleNewRange = this.handleNewRange.bind(this);
  }

  componentDidMount() {
    this.calculateTotalDuration();
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.ranges instanceof Array && (
        !(prevProps.ranges instanceof Array) ||
        (prevProps.ranges.length !== this.props.ranges.length)
      )
    ) {
      this.calculateTotalDuration();
    }
  }

  calculateTotalDuration() {
    const { ranges } = this.props;
    const totalDuration = moment.duration();

    if (ranges instanceof Array) {
      ranges.forEach((range) => {
        const to = moment(range.to, 'hh:mm');
        const from = moment(range.from, 'hh:mm');

        totalDuration.add(moment.duration(to.diff(from)));
      });
    }

    this.setState({
      totalDuration,
    });
  }

  handleComplete() {
    const { id } = this.props;

    this.props.onComplete(id);
  }

  handleNewRange(from, to) {
    const { id } = this.props;

    this.props.onNewRange(id, from, to);
  }

  render() {
    const { date, completed, ranges } = this.props;
    const { totalDuration } = this.state;

    const renderRanges = () => {
      if (ranges instanceof Array) {
        return ranges.map((range, index) => {
          const to = moment(`${date}${range.to}`, 'YYYY-MM-DDhh:mm');
          const from = moment(`${date}${range.from}`, 'YYYY-MM-DDhh:mm');

          return (
            <div key={index}>
              from {from.format('LT')} to {to.format('LT')}
            </div>
          );
        });
      }

      return (
        <div className="callout warning">
          No data yet. Maybe add some?
        </div>
      );
    };

    const renderControls = () => {
      if (!completed) {
        return <ItemControls onComplete={this.handleComplete} onNewRange={this.handleNewRange} />;
      }

      return (
        <div />
      );
    };

    return (
      <div className="callout clearfix">
        <div className="float-right">
          {totalDuration.hours()} h {totalDuration.minutes()} m
        </div>
        <p><b>{moment(date).format('dddd, MMMM Do YYYY')}</b></p>
        <div className="time-ranges">
          Time ranges:
          {renderRanges()}
        </div>
        {renderControls()}
      </div>
    );
  }
}

export const trackerItemProps = {
  id: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  completed: PropTypes.bool.isRequired,
  ranges: PropTypes.arrayOf(
    PropTypes.shape({
      from: PropTypes.string.isRequired,
      to: PropTypes.string.isRequired,
    }),
  ),
};

TrackerItem.propTypes = Object.assign({}, trackerItemProps, {
  onNewRange: PropTypes.func.isRequired,
  onComplete: PropTypes.func.isRequired,
});

export default TrackerItem;
