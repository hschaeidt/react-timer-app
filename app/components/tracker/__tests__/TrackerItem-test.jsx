import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import TrackerItem from '../TrackerItem';
import tracks from '../../../fixtures/tracks.json';

function setupWrapper(fixtureIndex) {
  const props = Object.assign({}, tracks[fixtureIndex], {
    onNewRange: jest.fn(),
    onComplete: jest.fn(),
  });

  const wrapper = mount(<TrackerItem {...props} />);

  return {
    wrapper,
    props,
  };
}

function setupTree(fixtureIndex) {
  const props = Object.assign({}, tracks[fixtureIndex], {
    onNewRange: jest.fn(),
    onComplete: jest.fn(),
  });

  const tree = renderer.create(<TrackerItem {...props} />);

  return {
    tree,
    props,
  };
}

describe('<TrackerItem />', () => {
  it('renders correctly', () => {
    const { tree } = setupTree(0);
    expect(tree).toMatchSnapshot();
  });

  it('should call handleComplete when lock button clicked', () => {
    const { wrapper, props } = setupWrapper(1);
    wrapper.find('#lock-button').simulate('click');
    expect(props.onComplete).toHaveBeenCalledWith(props.id);
  });

  it('should onNewRange prop on handling new range', () => {
    const { wrapper, props } = setupWrapper(1);
    wrapper.instance().handleNewRange('21:00', '22:00');
    expect(props.onNewRange).toHaveBeenCalledWith(tracks[1].id, '21:00', '22:00');
  });

  it('should update totalDuration when new range item added', () => {
    const { wrapper, props } = setupWrapper(1);
    const { ranges } = props;
    const instance = wrapper.instance();
    expect(instance.state.totalDuration.hours()).toBe(8);
    wrapper.setProps({
      ranges: [
        ...ranges,
        {
          from: '20:00',
          to: '21:00',
        },
      ],
    });
    expect(instance.state.totalDuration.hours()).toBe(9);
  });

  describe('render', () => {
    it('should render with controls if not locked', () => {
      const { tree } = setupTree(1);
      expect(tree).toMatchSnapshot();
    });

    it('should render without controls if locked', () => {
      const { tree } = setupTree(0);
      expect(tree).toMatchSnapshot();
    });

    it('should be able to render without time ranges', () => {
      const { tree } = setupTree(2);
      expect(tree).toMatchSnapshot();
    });
  });
});
