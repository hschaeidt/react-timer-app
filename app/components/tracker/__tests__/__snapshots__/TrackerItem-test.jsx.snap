exports[`<TrackerItem /> render should be able to render without time ranges 1`] = `
<div
  className="callout clearfix">
  <div
    className="float-right">
    0
     h 
    0
     m
  </div>
  <p>
    <b>
      Saturday, December 3rd 2016
    </b>
  </p>
  <div
    className="time-ranges">
    Time ranges:
    <div
      className="callout warning">
      No data yet. Maybe add some?
    </div>
  </div>
  <div
    className="button-group">
    <form
      id="handle-new-range-form"
      onSubmit={[Function]}>
      <input
        placeholder="--:--"
        type="time" />
      <input
        placeholder="--:--"
        type="time" />
      <button
        className="button primary">
        Add range
      </button>
    </form>
    <button
      className="button alert hollow"
      id="lock-button"
      onClick={[Function]}>
      Lock
    </button>
  </div>
</div>
`;

exports[`<TrackerItem /> render should render with controls if not locked 1`] = `
<div
  className="callout clearfix">
  <div
    className="float-right">
    8
     h 
    0
     m
  </div>
  <p>
    <b>
      Saturday, December 3rd 2016
    </b>
  </p>
  <div
    className="time-ranges">
    Time ranges:
    <div>
      from 
      10:00 AM
       to 
      1:00 PM
    </div>
    <div>
      from 
      2:00 PM
       to 
      7:00 PM
    </div>
  </div>
  <div
    className="button-group">
    <form
      id="handle-new-range-form"
      onSubmit={[Function]}>
      <input
        placeholder="--:--"
        type="time" />
      <input
        placeholder="--:--"
        type="time" />
      <button
        className="button primary">
        Add range
      </button>
    </form>
    <button
      className="button alert hollow"
      id="lock-button"
      onClick={[Function]}>
      Lock
    </button>
  </div>
</div>
`;

exports[`<TrackerItem /> render should render without controls if locked 1`] = `
<div
  className="callout clearfix">
  <div
    className="float-right">
    8
     h 
    4
     m
  </div>
  <p>
    <b>
      Friday, December 2nd 2016
    </b>
  </p>
  <div
    className="time-ranges">
    Time ranges:
    <div>
      from 
      8:57 AM
       to 
      1:01 PM
    </div>
    <div>
      from 
      2:00 PM
       to 
      6:00 PM
    </div>
  </div>
  <div />
</div>
`;

exports[`<TrackerItem /> renders correctly 1`] = `
<div
  className="callout clearfix">
  <div
    className="float-right">
    8
     h 
    4
     m
  </div>
  <p>
    <b>
      Friday, December 2nd 2016
    </b>
  </p>
  <div
    className="time-ranges">
    Time ranges:
    <div>
      from 
      8:57 AM
       to 
      1:01 PM
    </div>
    <div>
      from 
      2:00 PM
       to 
      6:00 PM
    </div>
  </div>
  <div />
</div>
`;
